py/builtinimport.c

```diff
@@ -347,128 +352,186 @@ STATIC void evaluate_relative_import(mp_int_t level, const char **module_name, s
-// full_mod_name:    The full absolute path to this module (e.g. "foo.bar.baz").
+// full_mod_name:    The full absolute path up to this level (e.g. "foo.bar.baz").
 // level_mod_name:   The final component of the path (e.g. "baz").
 // outer_module_obj: The parent module (we need to store this module as an
 //                   attribute on it) (or MP_OBJ_NULL for top-level).
-// path:             The filesystem path where we found the parent module
-//                   (or empty for a top level module).
 // override_main:    Whether to set the __name__ to "__main__" (and use __main__
 //                   for the actual path).
-STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name, mp_obj_t outer_module_obj, vstr_t *path, bool override_main) {
-    mp_import_stat_t stat = MP_IMPORT_STAT_NO_EXIST;
+STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name, mp_obj_t outer_module_obj, bool override_main) {
+    // Immediately return if the module at this level is already loaded.
+    mp_map_elem_t *elem;
 
-    // Exact-match of built-in (or already-loaded) takes priority.
-    mp_obj_t module_obj = mp_module_get_loaded_or_builtin(full_mod_name);
+    #if MICROPY_PY_SYS
+    // If sys.path is empty, the intention is to force using a built-in. This
+    // means we should also ignore any loaded modules with the same name
+    // which may have come from the filesystem.
+    size_t path_num;
+    mp_obj_t *path_items;
+    mp_obj_list_get(mp_sys_path, &path_num, &path_items);
+    if (path_num)
+    #endif
+    {
+        elem = mp_map_lookup(&MP_STATE_VM(mp_loaded_modules_dict).map, MP_OBJ_NEW_QSTR(full_mod_name), MP_MAP_LOOKUP);
+        if (elem) {
+            return elem->value;
+        }
+    }
 
-    // Even if we find the module, go through the motions of searching for it
-    // because we may actually be in the process of importing a sub-module.
-    // So we need to (re-)find the correct path to be finding the sub-module
-    // on the next iteration of process_import_at_level.
+    VSTR_FIXED(path, MICROPY_ALLOC_PATH_MAX);
+    mp_import_stat_t stat = MP_IMPORT_STAT_NO_EXIST;
+    mp_obj_t module_obj;
 
     if (outer_module_obj == MP_OBJ_NULL) {
         DEBUG_printf("Searching for top-level module\n");
 
+        // Always prefer an exact match for any u-prefixed built-in
+        // (or `micropython` or `sys`), and bypass the filesystem.
+        const char *level_mod_name_str = qstr_str(level_mod_name);
+        if (level_mod_name_str[0] == 'u' || level_mod_name == MP_QSTR_micropython) {
+            module_obj = mp_module_get_builtin(level_mod_name);
+            if (module_obj != MP_OBJ_NULL) {
+                return module_obj;
+            }
+        }
+        #if MICROPY_PY_SYS
+        // Never allow sys to be overridden from the filesystem. If weak links
+        // are disabled, then this also provides a default weak link so that
+        // `import sys` works.
```

Comment:
Note that CPython definitely handles `sys` differently because it doesn't have `sys.__file__`.